### ISSUE TITLE
contours of zero width can't be exported into a variable font

### ISSUE THREAD

**Issue by jansindl3r** [googlefonts/fontmake#572]
Hello, I have a master that has all points at the same x position. It cannot be exported as variable font with fontmake -o variable -m font.designspace. Source is in ufo. I am attaching demo that reproduces this. Can I turn off compatibility check and force the export? See letter A Thanks, Jan
[bug.zip](https://github.com/googlefonts/fontmake/files/3517705/bug.zip)

`--keep-direction` and `--keep-overlaps` didn't help
CLI:
`fontmake -o variable -m bug.designspace --verbose WARNING`
Error:
`WARNING:fontTools.varLib:glyph A has incompatible masters; skipping`

**Comment by jansindl3r** [comment]
What is uncompatible in this? 

```
  <outline>
    <contour>
      <point type="line" x="72" y="651" />
      <point type="line" x="72" y="101" />
      <point type="line" x="476" y="101" />
      <point type="line" x="476" y="651" />
    </contour>
  </outline>
```

vs


```
  <outline>
    <contour>
      <point type="line" x="0" y="651" />
      <point type="line" x="0" y="101" />
      <point type="line" x="0" y="101" />
      <point type="line" x="0" y="651" />
    </contour>
  </outline>
```

**Comment by madig** [comment]
Maybe the points
```
      <point type="line" x="0" y="101" />
      <point type="line" x="0" y="101" />
```
get merged into one point?

**Comment by jansindl3r** [comment]
Most likely, I dont want them to get merged, what could be done, please? 
I also believe that only overlapped starting point gets merged - will check tomorrow.

**Comment by jansindl3r** [comment]
this is what Glyphs are able to export into a variable font 

```
master 1
x:83.0 y:129.0 type:line
x:518.0 y:129.0 type:line
x:518.0 y:607.0 type:line
x:83.0 y:607.0 type:line
```

```
Master 2 (everything at the same position)
x:300.0 y:350.0 type:line
x:300.0 y:350.0 type:line
x:300.0 y:350.0 type:line
x:300.0 y:350.0 type:line
```

**Comment by jansindl3r** [comment]
same problem with arg -o ttf-interpolatable ...overlapped startpoint gets merged

**Comment by madig** [comment]
Hm. It werks if I do `fontmake.exe -m bug.designspace -o ttf --keep-overlaps` but fails with `ttf-interpolatable`.

**Comment by madig** [comment]
Tracking this down to `cu2qu.ufo.fonts_to_quadratic`...

**Comment by madig** [comment]
This seems to be a feature (?) of fontTools' `ReverseContourPen`. Making the https://github.com/fonttools/fonttools/blob/63fb3fb881440b636267dc43082aab134e40f8f9/Lib/fontTools/pens/reverseContourPen.py#L64 unconditionally true turns the input points

```
[
    (0, 651),
    (0, 101),
    (0, 101),
    (0, 651),
]
```
into
```
[
    (0, 651),
    (0, 651),
    (0, 101),
    (0, 101),
]
```

The associated cu2qu test case would be 
```Python
    def test_overlapping_points(self):
        glyph = Glyph()
        pen = glyph.getPointPen()
        pen.beginPath()
        pen.addPoint((0, 651), segmentType="line")
        pen.addPoint((0, 101), segmentType="line")
        pen.addPoint((0, 101), segmentType="line")
        pen.addPoint((0, 651), segmentType="line")
        pen.endPath()
        assert glyph_to_quadratic(glyph, reverse_direction=True)
        assert [(p.x, p.y) for p in glyph[0]] == [
            (0, 651),
            (0, 651),
            (0, 101),
            (0, 101),
        ]
```
which passes with the modification mentioned above. Not sure what to do about this one. @anthrotype

**Comment by jansindl3r** [comment]
@madig thanks for having look on that! I tried to track it, but I wasn't able to go that deep. The reordering is wrong as long as it happens per master, none of the apps out there is able to export my variable font correctly, only FontLab VI for Mac can do it. I don't know, probably they use their own version of FontTools. They have option to trun on and off these "smart features" that shift or remove points randomly

**Comment by jansindl3r** [comment]
@madig I have just checked and FontLab exports it correctly and is using the same `ReverseContourPen.py` as you pointed to. It must be somewhere else, then?

**Comment by anthrotype** [comment]
@jansindl3r I fixed the issue with duplicate points being dropped in fonttools PointToSegmentPen. The fix is available in the latest fontmake v2.0.2 (which requires fonttools v4.0.1).
Please let us know if you find other issues, thanks.

**Comment by jansindl3r** [comment]
@anthrotype I have just tested it on my font..

using command fontmake of version using indeed 2.0.2  in virtual enviroment is raising this error. 
`ERROR:cu2qu.ufo:Glyphs named 'A' have different number of segments` Some of the glyphs went though and are displayed correctly, so it is on good way! About half of the glyphs returned error. I am using this arguments command `fontmake -o variable --keep-overlaps -m bug.designspace`
I have attached source causing it, you find there one glyph that gets export and one that gets not. 

`ERROR:cu2qu.ufo:Glyphs named 'A' have different number of segments`

[bug.zip](https://github.com/googlefonts/fontmake/files/3601940/bug.zip)

**Comment by anthrotype** [comment]
@jansindl3r thanks for the files. I think I found the root of the issue. I will push a fix to cu2qu soon.
Nice rotation effect, btw 👍

**Comment by anthrotype** [comment]
@jansindl3r I have released [cu2qu 1.6.6](https://github.com/googlefonts/cu2qu/releases/tag/v1.6.6) with the fix and verified that your test font builds successfully with latest fontmake and updated cu2qu.

**Comment by jansindl3r** [comment]
@anthrotype thanks! It works in full version of the font as well.. I am very happy that I can start implementing fontmake into my workflow.. 
![rotor_better_op](https://user-images.githubusercontent.com/44162266/64875328-28645280-d64d-11e9-89ef-e45dc1123fef.gif)
